pkg/sql/insert.go

```diff
@@ -82,6 +88,73 @@ type insertRun struct {
 
 	// traceKV caches the current KV tracing flag.
 	traceKV bool
+
+	// regionLocalInfo handles erroring out the INSERT when the
+	// enforce_home_region setting is on.
+	regionLocalInfo regionLocalInfoType
+}
+
+// regionLocalInfoType contains common items needed for determining the home region
+// of an insert or update operation.
+type regionLocalInfoType struct {
+
+	// regionMustBeLocalColID indicates the id of the column which must use a
+	// home region matching the gateway region in a REGIONAL BY ROW table.
+	regionMustBeLocalColID descpb.ColumnID
+
+	// gatewayRegion is the string representation of the gateway region when
+	// regionMustBeLocalColID is non-zero.
+	gatewayRegion string
+
+	// colIDtoRowIndex is the map to use to decode regionMustBeLocalColID into
+	// an index into the Datums slice corresponding with the crdb_region column.
+	colIDtoRowIndex catalog.TableColMap
+}
+
+// setupEnforceHomeRegion sets regionMustBeLocalColID and the gatewayRegion name
+// if we're enforcing a home region for this insert run. The colIDtoRowIndex map
+// to use to when translating column id to row index is also set up.
+func (r *regionLocalInfoType) setupEnforceHomeRegion(
+	p *planner, table cat.Table, cols []catalog.Column, colIDtoRowIndex catalog.TableColMap,
+) {
+	if p.EnforceHomeRegion() {
+		if gatewayRegion, ok := p.EvalContext().Locality.Find("region"); ok {
+			if homeRegionColName, ok := table.HomeRegionColName(); ok {
+				for _, col := range cols {
+					if col.ColName() == tree.Name(homeRegionColName) {
+						r.regionMustBeLocalColID = col.GetID()
+						r.gatewayRegion = gatewayRegion
+						r.colIDtoRowIndex = colIDtoRowIndex
+						break
+					}
+				}
+			}
+		}
+	}
+}
+
+// checkHomeRegion errors out the insert or update if the enforce_home_region session setting is on and
+// the row's locality doesn't match the gateway region.
+func (r *regionLocalInfoType) checkHomeRegion(row tree.Datums) error {
+	if r.regionMustBeLocalColID != 0 {
+		if regionColIdx, ok := r.colIDtoRowIndex.Get(r.regionMustBeLocalColID); ok {
+			if regionEnum, regionColIsEnum := row[regionColIdx].(*tree.DEnum); regionColIsEnum {
+				if regionEnum.LogicalRep != r.gatewayRegion {
```

Comment:
You can reduce these deeply nested conditionals by inverting the condition and early-returning, which is more idiomatic in Go and our codebase, e.g.:

```go
if r.regionMustBeLocalColID == 0 {
  return nil
}
// ...
```

No need to change this now though, since it's a backport.